### PR TITLE
cms/396

### PR DESCRIPTION
**Pull request by danieljust**
added possibility for adding user info in db, after submitting regform

### CODE REVIEW

File: backend/app/controllers/registration.controller.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module.exports = (request, response) => {
+    response.render('pages/registration');
+};
```
Comment (mvahrameev) [inline backend/app/controllers/registration.controller.js]: Rename to `registration.js`

File: backend/public/css/login.css [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+.t-loginpage-wrapper {
+    display: table-cell;
+    margin: 24px;
+    white-space: normal;
+    vertical-align: middle;
+    min-width: 450px;
+}
+
+.t-loginform {
```
Comment (mvahrameev) [inline backend/public/css/login.css]: Where is this class used?

File: backend/routes/render.route.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module.exports = function (router) {
+    const register = require('../app/controllers/registration.controller');
+    const index = require('../app/controllers/index');
+    const userpage = require('../app/controllers/userpage.controller');
```
Comment (mvahrameev) [inline backend/routes/render.route.js]: Move decalrations of this varaibles to the beginning of the file

File: backend/app/views/pages/registration.pug [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        title Registration
+        link(rel='stylesheet', href='css/registration.css')
+        link(rel='stylesheet', href='css/t-futura.css')
+        script(src='js/jquery-3.2.0.js')
+        script(src='js/btnclk.js')
+    body.theme
+        .regpage-full
+            .regpage-wrapper
+                .regform
+                    form(name='form', action='', method='post',enctype='application/json')
```
Comment (IgorZiegler) [inline backend/app/views/pages/registration.pug]: Do you really need empty action?

File: backend/public/css/login.css [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+body {
+    background-color: #ef6b50;
+}
+
+.t-loginpage {
+    max-width: 470px;
+    margin: 0 auto;
+    padding: 0 0;
```
Comment (IgorZiegler) [inline backend/public/css/login.css]: `padding: 0;`

File: backend/app.js [hunk 6]
```diff
@@ -1,13 +1,23 @@
 const express = require('express');
+
```
Comment (maxmalov) [inline backend/app.js]: why is that needed?

File: backend/app.js [hunk 7]
```diff
@@ -1,13 +1,23 @@
...
 const router = require('./routes/router');
 
 const app = express();
 
+const bodyParser = require('body-parser');
```
Comment (maxmalov) [inline backend/app.js]: it is a good practice to place all module requires at the top of the module

File: backend/app.js [hunk 8]
```diff
@@ -1,13 +1,23 @@
...
+const mongoose = require('mongoose');
+
 app.set('views', './app/views');
 app.set('view engine', 'pug');
 app.use(express.static('public'));
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({extended: true}));
 app.use(router);
-
+mongoose.connect(dbUrl);
```
Comment (maxmalov) [inline backend/app.js]: you should consider move this logic into a separate service module from application setup

File: backend/app/views/pages/registration.pug [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+                        .regform__items
+                            .regform__item
+                                input#name.regform-input_custom(placeholder='Ваше имя', autocomplete='off', type='text',name="name")
```
Comment (maxmalov) [inline backend/app/views/pages/registration.pug]: I guess we should ban element identifiers from now

File: backend/package.json [hunk 10]
```diff
@@ -4,12 +4,19 @@
   "description": "",
   "main": "app.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "dev": "nodemon app.js"
```
Comment (maxmalov) [inline backend/package.json]: it is a good practice to name this script `start`

File: backend/package.json [hunk 11]
```diff
@@ -4,12 +4,19 @@
...
   "license": "ISC",
   "dependencies": {
+    "body-parser": "^1.17.1",
     "express": "^4.15.2",
+    "mongodb": "^2.2.25",
+    "mongoose": "^4.9.1",
     "pug": "^2.0.0-beta11"
+  },
+  "devDependencies": {
+    "nodemon": "^1.11.0"
```
Comment (maxmalov) [inline backend/package.json]: :+1:

File: backend/public/css/login.css [hunk 12]
```diff
@@ -0,0 +1,148 @@
+html, body {
+    margin: 0;
+    height: 100%;
+    overflow-x: hidden;
+}
+
+input, textarea, select, button, body {
+    font-family: 'tfutura', Arial, sans-serif;
```
Comment (maxmalov) [inline backend/public/css/login.css]: font settings should be at the top level

File: backend/public/css/registration.css [hunk 13]
```diff
@@ -0,0 +1,202 @@
+html, body {
```
Comment (maxmalov) [inline backend/public/css/registration.css]: looks like a copypaste to me, consider extracting common css into a separate stylesheet

File: backend/public/css/t-futura.css [hunk 14]
```diff
@@ -0,0 +1,30 @@
+@font-face {
+    font-family: 'tfutura';
```
Comment (maxmalov) [inline backend/public/css/t-futura.css]: We're not sure this is an open font, so there might be licensing issues, since this repo is public. Please replace it with open fonts from google api for ex

File: backend/public/js/btnclk.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    if (typeof data.redirect == 'string') {
+        window.location = data.redirect
+    }
+}
+
+function getFormData($form) {
+    const unindexed_array = $form.serializeArray();
+    const indexed_array = {};
+    $.map(unindexed_array, function (n, i) {
+        indexed_array[n['name']] = n['value'];
```
Comment (maxmalov) [inline backend/public/js/btnclk.js]: Why do you use bracket style notation to access object properties?

File: backend/public/js/btnclk.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+function onAjaxSuccess(data, textStatus, jqXHR) {
+    if (typeof data.redirect == 'string') {
+        window.location = data.redirect
+    }
+}
+
+function getFormData($form) {
+    const unindexed_array = $form.serializeArray();
+    const indexed_array = {};
+    $.map(unindexed_array, function (n, i) {
```
Comment (maxmalov) [inline backend/public/js/btnclk.js]: there is a map available in the Array api

File: backend/public/js/btnclk.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+});
+
+function onAjaxSuccess(data, textStatus, jqXHR) {
+    if (typeof data.redirect == 'string') {
+        window.location = data.redirect
+    }
+}
+
+function getFormData($form) {
+    const unindexed_array = $form.serializeArray();
```
Comment (maxmalov) [inline backend/public/js/btnclk.js]: you should use the same naming style across all application, no mix of snake and camel cases

File: backend/routes/user.route.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
+module.exports = function (app) {
+    const mongoose = require('mongoose');
+    const userSchema = mongoose.Schema({
```
Comment (maxmalov) [inline backend/routes/user.route.js]: Do you even decompose? What if you'll need this model somewhere else in the application?

File: backend/routes/user.route.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    });
+    const Person = mongoose.model('users', userSchema);
+
+    app.post('/registration', (req, res) => {
+            const samplePerson = new Person({
+                name: req.body.name,
+                email: req.body.email,
+                password: req.body.pwd
+            });
+            samplePerson.save(function (err, samplePerson) {
```
Comment (maxmalov) [inline backend/routes/user.route.js]: Let say after a month we'll decide to drop mongodb and use postgresql. With that knowledge make sure that this code are not aware of concrete database

File: backend/package.json (hunk 11 above)
Comment (danieljust) [inline backend/package.json]: thanks :)

File: backend/routes/user.js [hunk 20]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require('mongoose');
+const dbService = require('../app/dao/userDAOImpl/mongoUserMethods');
+const userLoginSchema = require('../app/schemas/user').userLoginSchema;
+
+module.exports = function (app) {
+    const Person = mongoose.model('users', userLoginSchema);
```
Comment (maxmalov) [inline backend/routes/user.js]: You should move all mongoose related logic into services, even model references should be behind your abstractions. I'll provide a brief overview of the structure for DAO related service:

- You should have a separate module for each model. In your case it can be covered by mongoose models. This will be concrete implementations for mongodb backend.
- You should have a main facade for user DAO service, which will delegate all calls to mongoose models. This facade shouldn't have abstraction leaks to mongoose models. In this example you manually create mongoose model and specify `mongoUserMethods` in require call
- Finally you should have a layer called use cases. Ideally each server endpoint is an use case, so endpoint handlers just simply call these use cases. In other words use cases level is a border btw your app and express itself. Use cases operates directly with DAO related services, so they don't know anything about the db backend.

File: backend/routes/user.js [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+    app.post('/registration', (req, res) => {
```
Comment (maxmalov) [inline backend/routes/user.js]: you should also move this handler into a controller

File: backend/app/views/pages/registration.pug [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+                                input.regform-input_custom(placeholder='Ваше имя', autocomplete='off', type='text',name="name")
+                            .regform__item
+                                input.regform-input_custom(placeholder='Электронная почта', autocomplete='off', type='text',name="email")
+                            .regform__item
+                                input.regform-input_custom(placeholder='Пароль', autocomplete='off', type='password',name="pwd")
+                            .text-center.item
+                                button.regform-btn(type='submit')
+                                    | Зарегистрироваться
+                    .regform-link__terms
+                        | Нажимая кнопку «Зарегистрироваться», Вы принимаете условия&nbsp
```
Comment (102) [inline backend/app/views/pages/registration.pug]: This probably should have semicolon at the end: `&nbsp;`

File: backend/app/dao/services/dbService.js [hunk 23]
```diff
@@ -0,0 +1,7 @@
+const mongoUserDAO = require('./mongoUser');
+
+module.exports = {
+    save: function (requestedObject, res) {
```
Comment (maxmalov) [inline backend/app/dao/services/dbService.js]: why do you need the `res` here?

File: backend/app/dao/useCases.js [hunk 24]
```diff
@@ -0,0 +1,12 @@
+const dbService = require('./services/dbService');
```
Comment (maxmalov) [inline backend/app/dao/useCases.js]: use cases should live outside of `DAO` related code

Comment (maxmalov) [inline backend/app/dao/useCases.js]: `useCases.js` is a very common name for this type of use case

File: backend/app/dao/services/dbService.js [hunk 25]
```diff
@@ -0,0 +1,7 @@
+const mongoUserDAO = require('./mongoUser');
```
Comment (maxmalov) [inline backend/app/dao/services/dbService.js]: `dbService` is a very common name for this module

File: backend/app.js [hunk 26]
```diff
@@ -1,13 +1,16 @@
...
 const app = express();
 
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({extended: true}));
 app.set('views', './app/views');
 app.set('view engine', 'pug');
 app.use(express.static('public'));
 app.use(router);
-
+db.start();
```
Comment (maxmalov) [inline backend/app.js]: express application should be unaware of db related stuff
